### PR TITLE
Pagination: add tests for component

### PR DESCRIPTION
**Pull request by nathanrodman**
**What this PR does / why we need it**:

Adds some tests to the pagination component. Since there's a lot of lengthy logic in the reducer it felt better to have some tests that covered all the use cases. This was some of the feedback addressed in https://github.com/grafana/grafana/pull/30151

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

### CODE REVIEW

File: packages/grafana-ui/src/components/Pagination/Pagination.test.tsx [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+};
+const getNextButton = async () => {
+  return await screen.getByLabelText('next');
+};
+const getPreviousButton = async () => {
+  return await screen.getByLabelText('previous');
+};
+const getAllPaginationElements = async () => {
+  return await screen.findAllByRole('listitem');
+};
```
Comment (Clarity-89) [inline packages/grafana-ui/src/components/Pagination/Pagination.test.tsx:17]:
```suggestion
const getPageButton = (page: number) => {
  return screen.queryByText(`${page}`);
};
const getNextButton = () => {
  return screen.getByLabelText('next');
};
const getPreviousButton = () => {
  return screen.getByLabelText('previous');
};
const getAllPaginationElements = () => {
  return screen.findAllByRole('listitem');
};
```
No need to use `asyn/await` if the only thing the function does is return a promise. Plus, only `find*` queries return promises.

File: packages/grafana-ui/src/components/Pagination/Pagination.test.tsx [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+describe('Pagination', () => {
+  let currentPage = 1;
+  const onNavigate = (page: number) => {
+    currentPage = page;
+  };
+  const MAX_ELEMENTS = 11;
+  it('renders the correct number of pages', async () => {
+    render(<Pagination numberOfPages={3} currentPage={currentPage} onNavigate={onNavigate} />);
+    const listItems = await getAllPaginationElements();
+    expect(listItems.length).toEqual(5);
```
Comment (Clarity-89) [inline packages/grafana-ui/src/components/Pagination/Pagination.test.tsx:28]:
```suggestion
    expect(listItems).toHaveLength(5);
```

File: packages/grafana-ui/src/components/Pagination/Pagination.test.tsx [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+  });
+
+  it('renders the condensed version for a large number of pages', async () => {
+    render(<Pagination numberOfPages={100} currentPage={currentPage} onNavigate={onNavigate} />);
+    const listItems = await getAllPaginationElements();
+    expect(listItems.length).toEqual(MAX_ELEMENTS);
```
Comment (Clarity-89) [inline packages/grafana-ui/src/components/Pagination/Pagination.test.tsx:34]:
```suggestion
    expect(listItems).toHaveLength(MAX_ELEMENTS);
```

File: packages/grafana-ui/src/components/Pagination/Pagination.test.tsx [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+  });
+
+  it('increments the page', async () => {
+    currentPage = 9;
+    render(<Pagination numberOfPages={13} currentPage={currentPage} onNavigate={onNavigate} />);
+    const nextButton = await getNextButton();
```
Comment (Clarity-89) [inline packages/grafana-ui/src/components/Pagination/Pagination.test.tsx:40]:
```suggestion
    const nextButton = getNextButton();
```
`get*` queries don't return a promise.

File: packages/grafana-ui/src/components/Pagination/Pagination.test.tsx [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+
+    await userEvent.click(nextButton);
```
Comment (Clarity-89) [inline packages/grafana-ui/src/components/Pagination/Pagination.test.tsx:42]:
```suggestion
   userEvent.click(nextButton);
```

File: packages/grafana-ui/src/components/Pagination/Pagination.test.tsx [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+    expect(currentPage).toEqual(10);
+  });
+
+  it('decrements the page', async () => {
+    currentPage = 9;
+    render(<Pagination numberOfPages={13} currentPage={currentPage} onNavigate={onNavigate} />);
+    const previousButton = await getPreviousButton();
```
Comment (Clarity-89) [inline packages/grafana-ui/src/components/Pagination/Pagination.test.tsx:49]:
```suggestion
    const previousButton = getPreviousButton();
```

File: packages/grafana-ui/src/components/Pagination/Pagination.test.tsx [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+
+    await userEvent.click(previousButton);
```
Comment (Clarity-89) [inline packages/grafana-ui/src/components/Pagination/Pagination.test.tsx:51]:
```suggestion
   userEvent.click(previousButton);
```

File: packages/grafana-ui/src/components/Pagination/Pagination.test.tsx [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+
+  it('navigates to a specific page when clicking', async () => {
+    currentPage = 1;
+    render(<Pagination numberOfPages={25} currentPage={currentPage} onNavigate={onNavigate} />);
+    const buttonSeven = await screen.findByText('7');
+    userEvent.click(buttonSeven);
+    expect(currentPage).toEqual(7);
+  });
+
+  it('does not render condensed pages', async () => {
```
Comment (Clarity-89) [inline packages/grafana-ui/src/components/Pagination/Pagination.test.tsx:63]:
```suggestion
  it('does not render condensed pages', () => {
```

File: packages/grafana-ui/src/components/Pagination/Pagination.test.tsx [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+    currentPage = 1;
+    render(<Pagination numberOfPages={25} currentPage={currentPage} onNavigate={onNavigate} />);
+
+    const button = await getPageButton(24);
```
Comment (Clarity-89) [inline packages/grafana-ui/src/components/Pagination/Pagination.test.tsx:67]:
```suggestion
    const button = getPageButton(24);
```

File: packages/grafana-ui/src/components/Pagination/Pagination.test.tsx [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+    expect(button).toBeNull();
+  });
+
+  it('always renders the same number of elements for large number of pages', async () => {
+    currentPage = 60;
+    const { rerender } = render(<Pagination numberOfPages={100} currentPage={currentPage} onNavigate={onNavigate} />);
+
+    const pageListItems = await getAllPaginationElements();
+    expect(pageListItems.length).toEqual(MAX_ELEMENTS);
```
Comment (Clarity-89) [inline packages/grafana-ui/src/components/Pagination/Pagination.test.tsx:76]:
```suggestion
    expect(pageListItems).toHaveLength(MAX_ELEMENTS);
```

File: packages/grafana-ui/src/components/Pagination/Pagination.test.tsx [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+
+    rerender(<Pagination numberOfPages={50} currentPage={50} onNavigate={onNavigate} />);
+
+    const pages = await getAllPaginationElements();
+    expect(pages.length).toEqual(MAX_ELEMENTS);
```
Comment (Clarity-89) [inline packages/grafana-ui/src/components/Pagination/Pagination.test.tsx:81]:
```suggestion
    expect(pages).toHaveLength(MAX_ELEMENTS);
```

File: packages/grafana-ui/src/components/Pagination/Pagination.test.tsx [hunk 12]
```diff
@@ -0,0 +1,131 @@
...
+  });
+
+  it('renders the two next and two previous pages in condensed view', async () => {
```
Comment (Clarity-89) [inline packages/grafana-ui/src/components/Pagination/Pagination.test.tsx:84]:
```suggestion
  it('renders the two next and two previous pages in condensed view', () => {
```

File: packages/grafana-ui/src/components/Pagination/Pagination.test.tsx [hunk 13]
```diff
@@ -0,0 +1,131 @@
...
+    currentPage = 12;
+    const numberOfPages = 25;
+
+    render(<Pagination numberOfPages={numberOfPages} currentPage={currentPage} onNavigate={onNavigate} />);
+    const lowerMiddleButton = await getPageButton(currentPage - 2);
+    const upperMiddleButton = await getPageButton(currentPage + 2);
+    const firstPage = await getPageButton(1);
+    const lastPage = await getPageButton(numberOfPages);
```
Comment (Clarity-89) [inline packages/grafana-ui/src/components/Pagination/Pagination.test.tsx:92]:
```suggestion
    const lowerMiddleButton = getPageButton(currentPage - 2);
    const upperMiddleButton = getPageButton(currentPage + 2);
    const firstPage = getPageButton(1);
    const lastPage = getPageButton(numberOfPages);
```

File: packages/grafana-ui/src/components/Pagination/Pagination.test.tsx [hunk 14]
```diff
@@ -0,0 +1,131 @@
...
+  });
+
+  it('shows only relevant pages near beginning of page list', async () => {
+    currentPage = 1;
+    const numberOfPages = 12;
+    render(<Pagination numberOfPages={numberOfPages} currentPage={currentPage} onNavigate={onNavigate} />);
+
+    const seventhPage = await getPageButton(7);
+    const eighthPage = await getPageButton(8);
+    const lastPage = await getPageButton(numberOfPages);
```
Comment (Clarity-89) [inline packages/grafana-ui/src/components/Pagination/Pagination.test.tsx:107]:
```suggestion
    const seventhPage = getPageButton(7);
    const eighthPage = getPageButton(8);
    const lastPage = getPageButton(numberOfPages);
```

File: packages/grafana-ui/src/components/Pagination/Pagination.test.tsx [hunk 15]
```diff
@@ -0,0 +1,131 @@
...
+    const listItems = await getAllPaginationElements();
+
+    expect(seventhPage).not.toBeNull();
+    expect(eighthPage).toBeNull();
+    expect(lastPage).not.toBeNull();
+    expect(listItems.length).toEqual(MAX_ELEMENTS);
```
Comment (Clarity-89) [inline packages/grafana-ui/src/components/Pagination/Pagination.test.tsx:113]:
```suggestion
    expect(listItems).toHaveLength(MAX_ELEMENTS);
```

File: packages/grafana-ui/src/components/Pagination/Pagination.test.tsx [hunk 16]
```diff
@@ -0,0 +1,131 @@
...
+  });
+
+  it('shows only relevant pages near end of page list', async () => {
+    currentPage = 11;
+    const numberOfPages = 12;
+    render(<Pagination numberOfPages={numberOfPages} currentPage={currentPage} onNavigate={onNavigate} />);
+
+    const fifthPage = await getPageButton(5);
+    const sixthPage = await getPageButton(6);
+    const firstPage = await getPageButton(1);
```
Comment (Clarity-89) [inline packages/grafana-ui/src/components/Pagination/Pagination.test.tsx:123]:
```suggestion
    const fifthPage = getPageButton(5);
    const sixthPage = getPageButton(6);
    const firstPage = getPageButton(1);
```

**Review by Clarity-89 — COMMENTED** [review]
Looks good! Lots of async/await calls are redundant since `get*` queries do not return promises.

**Review by hugohaggmark — COMMENTED** [review]
Great, that you added these tests. Let me know if you need any help to fix the comments from @Clarity-89

**Pull request closed without merging**